File: src/main/java/me/partlysanestudios/partlysaneskies/gui/components/PSSItemRender.java
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//

package me.partlysanestudios.partlysaneskies.gui.components;

import gg.essential.elementa.UIComponent;
import gg.essential.elementa.constraints.PixelConstraint;
import gg.essential.universal.UMatrixStack;
import me.partlysanestudios.partlysaneskies.PartlySaneSkies;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.entity.RenderItem;
import net.minecraft.item.ItemStack;

public class PSSItemRender extends UIComponent {
    ItemStack item;
    float itemScale = 1;

    public PSSItemRender(ItemStack item) {
        this.item = item;
    }

    @Override
    public void draw(UMatrixStack matrixStack) {
        beforeDrawCompat(matrixStack);
        super.draw(matrixStack);
        drawItemStack(item, Math.round(this.getLeft()), Math.round(this.getTop()), getComponentName());
        super.afterDraw(matrixStack);
    }

    private void drawItemStack(ItemStack stack, int x, int y, String altText) {
        RenderItem itemRenderer = PartlySaneSkies.minecraft.getRenderItem();

        GlStateManager.pushMatrix();
        GlStateManager.scale(itemScale, itemScale, 1);
        itemRenderer.zLevel = 200f;
        net.minecraft.client.gui.FontRenderer font = null;
        if (stack != null)
            font = stack.getItem().getFontRenderer(stack);
        if (font == null)
            font = PartlySaneSkies.minecraft.fontRendererObj;
        itemRenderer.renderItemAndEffectIntoGUI(stack, Math.round(x / itemScale), Math.round(y / itemScale));
        GlStateManager.popMatrix();

    }

    public UIComponent setItemScale(PixelConstraint constraint) {
        this.itemScale = constraint.getValue();
        return this;
    }
}

File: src/main/kotlin/me/partlysanestudios/partlysaneskies/render/gui/components/PSSItemRender.kt
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//

package me.partlysanestudios.partlysaneskies.render.gui.components

import gg.essential.elementa.UIComponent
import gg.essential.elementa.constraints.PixelConstraint
import gg.essential.elementa.constraints.WidthConstraint
import gg.essential.elementa.dsl.pixels
import gg.essential.universal.UMatrixStack
import me.partlysanestudios.partlysaneskies.PartlySaneSkies.Companion.minecraft
import net.minecraft.client.gui.FontRenderer
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.item.ItemStack

class PSSItemRender(
    var item: ItemStack?,
    private val autoScaleWidth: Boolean = false,
) : UIComponent() {
    private var itemScale = 1f

    override fun draw(matrixStack: UMatrixStack) {
        beforeDrawCompat(matrixStack)
        super.draw(matrixStack)
        drawItemStack(item, Math.round(getLeft()), Math.round(getTop()), componentName)
        super.afterDraw(matrixStack)
    }

    private fun drawItemStack(
        stack: ItemStack?,
        x: Int,
        y: Int,
        altText: String,
    ) {
        if (autoScaleWidth) {
            this.constraints.height = this.constraints.width.cachedValue.pixels
            setScaleBasedOnWidth(this.constraints.width)
        }
        val itemRenderer = minecraft.renderItem
        GlStateManager.pushMatrix()
        GlStateManager.scale(itemScale, itemScale, 1f)
        itemRenderer.zLevel = 200f
        var font: FontRenderer? = null
        if (stack != null) font = stack.item.getFontRenderer(stack)
        if (font == null) font = minecraft.fontRendererObj
        itemRenderer.renderItemAndEffectIntoGUI(stack, Math.round(x / itemScale), Math.round(y / itemScale))
        GlStateManager.popMatrix()
    }

    fun setItem(item: ItemStack): PSSItemRender {
        this.item = item
        return this
    }

    fun setScaleBasedOnWidth(pixelConstraint: WidthConstraint): UIComponent {
        setItemScale((pixelConstraint.cachedValue / 16).pixels)
        return this
    }

    fun setItemScale(constraint: PixelConstraint): UIComponent {
        itemScale = constraint.value
        return this
    }
}
